[QoI] Diagnose initializers written as typed patterns (SR-1461)

jtbandes opened the pull request:
#### What's in this pull request?
Diagnose and offer to replace `var x: [Int]()` with `var x = [Int]()`.

```
error: unexpected initializer in pattern; did you mean to use '='?
var x: [Int]()
     ~ ~~~~~^~
      =
```

#### Resolved bug number: [SR-1461](https://bugs.swift.org/browse/SR-1461)
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

A smoke test on macOS does the following:

1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:

1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

jtbandes commented:
@swift-ci Please test

jtbandes commented:
@swift-ci Please test

jtbandes commented:
@swift-ci Please smoke test OS X platform

rintaro commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,14 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+          }
+          diagnose(lParenLoc, diag::initializer_as_typed_pattern)
+          .highlightChars(Ty.get()->getStartLoc(),
+                          rParenLoc.getAdvancedLoc(1))
```
Why not `.highlight({Ty.get()->getStartLoc(), rParenLoc})`

jtbandes replied:
Hmm, I guess I was wrong about the need to use getAdvancedLoc to highlight all the way to the closing paren. Will change.

rintaro commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,14 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+        }
+      }
+    }
+    else
```
Nitpick: `} else`

jtbandes replied:
I'll add another `{}` while I'm at it 😉

jtbandes commented:
@swift-ci Please smoke test OS X platform

tkremenek commented:
@swift-ci test macOS

jtbandes commented:
@tkremenek seems like something is broken here, at least regarding GitHub integration. I also notice it says "you're not authorized to merge this pull request".

tkremenek commented:
@jtbandes You cannot merge the PR because testing is still "waiting for status to be reported".  The problem we've seen recently is sometimes GitHub is not getting the notification properly from CI that a build has kicked off, even when it has.

tkremenek commented:
@swift-ci test macOS

tkremenek commented:
@jtbandes CI is now responding.  Waiting for tests.

jtbandes commented:
Even last night, when the tests had passed and github was aware of it, it still wouldn't have let me merge.

tkremenek commented:
@jtbandes That was for a "smoke test"; the requirements of the branch changed last night to "full test" since more people were using that and the CI seemed to be keeping up.  There was possibly a race between when the PR initiated CI and when the branch requirements changed.

jtbandes commented:
Is it supposed to take >3 hours? 😅

jtbandes commented:
Now this could use an actual review, perhaps @DougGregor or @lattner?

slavapestov commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+            replacement = "=";
+          }
+          diagnose(lParenLoc, diag::initializer_as_typed_pattern)
+          .highlight({Ty.get()->getStartLoc(), rParenLoc})
```
Indentation is wrong here

slavapestov commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+                                            trailingClosure);
+        if (status.isSuccess()) {
+          // Suggest replacing : with = (with proper whitespace).
+          llvm::StringRef replacement = " =";
```
I think its more idiomatic to use the methods on Lexer instead of extracting the raw source as a string.

jtbandes replied:
@slavapestov Which Lexer methods do you mean? Isn't the lexer mostly unaware of whitespace?

I modeled this off https://github.com/apple/swift/blob/d3d2b0afd7e33e14f46c8b772fae965bdf38aa33/lib/Sema/CSDiag.cpp#L5616 and https://github.com/apple/swift/blob/d3d2b0afd7e33e14f46c8b772fae965bdf38aa33/lib/AST/DiagnosticEngine.cpp#L150

rintaro replied:
`Lexer::getLocForEndOfToken(SourceMgr, result.get().getEndLoc()) != colonLoc`
should work.
I don't know which is better.

jtbandes replied:
I think I like the extractText version better, but happy to hear opinions from others who work with this code regularly.

jrose-apple replied:
I think @rintaro and @slavapestov both qualify for that by now. Besides, what if the previous character is a tab?

jrose-apple replied:
Nitpick: no need for `llvm::` on StringRef.

jtbandes commented:
@swift-ci test macOS

jtbandes commented:
@swift-ci test macOS

jrose-apple commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,45 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+          diagnose(lParenLoc, diag::initializer_as_typed_pattern)
+            .highlight({Ty.get()->getStartLoc(), rParenLoc})
+            .fixItReplace(colonLoc, replacement);
+          result.setIsParseError();
```
I'm a little unhappy that the parsed expr doesn't end up *anywhere.* That will probably break syntax highlighting and code completion. Is it possible to put it into the AST somewhere?

jtbandes replied:
What do you recommend? Might be able to hoist this into the caller, so if after parsing the pattern we find it's actually an initializer, we can parse it as such. It looked much easier the way it is when I implemented it, but I'll take another look later today.

jrose-apple replied:
Hm. @benlangmuir, @nkcsgexi, any ideas?

benlangmuir replied:
Hrm. I take it Swift's lexer isn't fancy enough to be able to modify the token stream?  If we backtracked and replaced the `:` with `=' the caller would probably recover very well.

Short of that, yeah trying to handle more of this in the caller might be workable...

nkcsgexi replied:
Can we at least preserve the TypeRepr, maybe parsed like `var x: [Int]`

nkcsgexi replied:
oh, my bad. You do preserve the TypeRepr. For the expressions in call arg, I've no other suggestions except hoisting to the caller as another expression.

jtbandes replied:
Hm, I guess there is a question of whether we want PatternBindingDecl parsing to generally recognize incorrect `var x: «expr»` and treat it as `var x = «expr»` instead.

(The fix as written is pretty simple because it only handles exprs that look like initializer calls. It handles `var x: [Int]()`, but it won't handle `var x: Unmanaged.fromOpaque(y)` or `var x: MemoryLayout<Int>.stride`, for instance.)

In general, if the user made a typo and wrote `:` instead of `=`, the parser might only notice a problem after *successfully* parsing a typed pattern binding. So, unless we want to try parsing `var x = «expr»` *first* every time we look at a pattern (even if we see a `:`), I think this would need to be hoisted even further out than the caller of `parseTypedPattern()`, because it needs to invoke all the logic in `parseExpr`/`parseExprPostfix`, and it might need to do so even if parseTypedPattern succeeded.

Which is to say: the fix as written is pretty simple and self-contained. After reading some more of the parser implementation, I'm inclined to think producing better results than this would be considerably more involved. Feel free to prove me wrong, if you can think of a clean way to do it!

nkcsgexi replied:
I like the idea of recognizing `var x : <<expr>>` as `var x = <<expr>>`. However, there may be other problems by doing so; for instance, if there is a function called `getEqualPos()` in the AST node, what will it return? a invalid position or the position of `:` instead.

The fixit per se is for sure an improvement! Thanks for working on that. Preserving erroneous AST nodes and providing insight of it is ideal for tooling but I have to say it is not easy in this case.

jrose-apple commented on test/Parse/diagnose_initializer_as_typed_pattern.swift:
```diff
@@ -0,0 +1,10 @@
+
+var x: SR1461(), y: Int  // expected-error{{unexpected initializer in pattern; did you mean to use '='?}} {{6-7= =}}
+
+var z : SR1461(), w: Int  // expected-error{{unexpected initializer in pattern; did you mean to use '='?}} {{7-8==}}
```
Please include the following tests too:

```swift
let a: X(x)
let a: X(x, y)
let a: X() { abc }
let a: X(x) { abc }
let a: X(x, y) { abc }

jrose-apple commented:
This is a great fix to have.

jtbandes commented:
@jrose-apple Thanks for the prompting on unit tests — that led me to discover trailing closures weren't handled properly at the top level, so I added a ContextChange.

I also cleaned up the logic for choosing a replacement string, so the fix-it won't produce invalid expressions like `var x =[Int]()`.

@swift-ci please smoke test macOS

jtbandes commented:
Why does GitHub hate to recognize my test runs? 😕
@swift-ci smoke test

jrose-apple commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,65 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+          // Suggest replacing ':' with '=' (ensuring proper whitespace).
+          
+          bool needSpaceBefore =
+            SourceMgr.getByteDistance(pastEndOfPrevLoc, colonLoc) == 0;
```
This can just use `==`.

jrose-apple replied:
This can just use `==`.

jrose-apple commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,65 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+          char chars[]  = " = ";
+          char *start   = &chars[0] + (needSpaceBefore ? 0 : 1);
+          char *pastEnd = &chars[3] - (needSpaceAfter  ? 0 : 1);
+          StringRef replacement(start, pastEnd - start);
```
I have a mild style preference for making the StringRef first and then using `drop_front`/`drop_back`.

jtbandes commented:
@jrose-apple done!
@swift-ci smoke test

jtbandes commented:
friendly bump @jrose-apple @DougGregor

jrose-apple commented:
I'm happy if @nkcsgexi and @rintaro are happy with it.

DougGregor commented:
LGTM!

rintaro commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -786,15 +786,63 @@ ParserResult<Pattern> Parser::parseTypedPattern() {
+          
+          bool needSpaceBefore = (pastEndOfPrevLoc == colonLoc);
+          bool needSpaceAfter =
+            SourceMgr.getByteDistance(colonLoc, startOfNextLoc) <= 1;
```
@jtbandes 
Sorry, this didn't work when a comment involves:
```swift
let i:/* comment */Int(42)
let i =/* comment */Int(42) // As of now, we can't parse this.
```
How do you think about #4499?
